### PR TITLE
[SPARK-12438][SQL] Add SQLUserDefinedType support for encoder

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-12438

ScalaReflection lacks the support of SQLUserDefinedType. We should add it.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @marmbrus @davies

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 1]
```diff
@@ -89,6 +94,23 @@ class UserDefinedTypeSuite extends QueryTest with SharedSQLContext with ParquetT
...
+  test("user type with ScalaReflection") {
+    val points = Seq(
+      MyLabeledPoint(1.0, new MyDenseVector(Array(0.1, 1.0))),
+      MyLabeledPoint(0.0, new MyDenseVector(Array(0.2, 2.0))))
+
+    val schema = ScalaReflection.schemaFor[MyLabeledPoint].dataType.asInstanceOf[StructType]
+    val attributeSeq = schema.toAttributes
+
+    val pointEncoder = encoderFor[MyLabeledPoint]
+    val unsafeRows = points.map(pointEncoder.toRow(_).copy())
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: can we also test `encoder.fromRow`?

we can just create a `MyLabelPoint` and encode it to `InternalRow` and decode it back by encoder, and check if the decoded `MyLabelPoint` is same with the original one.

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: I added the test now. By doing this, I also found that we need to add UserDefinedType to Cast to make it work.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -81,6 +81,9 @@ object Cast {
                 toField.nullable)
         }
 
+    case (udt1: UserDefinedType[_], udt2: UserDefinedType[_]) if udt1.userClass == udt2.userClass =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: I think this worth another JIRA.

Is `ScalaRelfection` the only place that may use UDT in `Cast`?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: I think so, as it is a special use case. I will open another JIRA for it.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: This fixing is submitted as pr #10410.

**Comment by marmbrus** [comment]
I think we need more design work before we implement this.  I'm not a huge fan of the current UDT API and I think we will replace it in Spark 2.0.

**Comment by viirya** [comment]
That is fine. But we will leave it broken until Spark2.0?

**Comment by marmbrus** [comment]
We can revisit this if we do a 1.7 release, but I don't think that is going to happen.

**Comment by viirya** [comment]
Personally, I would like to see this to be merged because it blocks my another PR #10283. Can we merge this and go back to modify this part when we revamp UDT API in the future? But of course it depends on you. Thanks.

**Comment by marmbrus** [comment]
Okay, I'm fine with merging this now to unblock other progress.  However, I think we should move the test to be with the other encoder tests.

**Comment by viirya** [comment]
@marmbrus Thanks. However, to move the test means that we need to move the UDT classes to encoder tests. I think it might duplicate the codes. Is it ok for you?

**Comment by marmbrus** [comment]
Do you mean just the example UDT `MyLabeledPoint`?  I'm fine with duplicating that or moving it entirely.  If UDTs are defined in catalyst we should also have unit tests for them there.

**Comment by viirya** [comment]
@marmbrus Not only MyLabeledPoint, which is just a case class. Also including UDT MyDenseVectorUDT and MyDenseVector. I think it might duplicate too many lines of codes from UserDefinedTypeSuite to encoder test suite like ExpressionEncoderSuite.

**Comment by cloud-fan** [comment]
We alread have UDT in [`RowEncoderSuite`](https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala#L28-L76), can we move it to a seperate file, and share it within catalyst tests?

**Comment by viirya** [comment]
@cloud-fan I just think the same thing. I will try it.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 3]
```diff
@@ -94,6 +99,9 @@ class UserDefinedTypeSuite extends QueryTest with SharedSQLContext with ParquetT
     assert(featuresArrays.contains(new MyDenseVector(Array(0.2, 2.0))))
   }
 
+  private val outers: ConcurrentMap[String, AnyRef] = new MapMaker().weakValues().makeMap()
+  outers.put(getClass.getName, this)
+
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: These changes are not needed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala [hunk 4]
```diff
@@ -242,6 +242,16 @@ class ExpressionEncoderSuite extends SparkFunSuite {
...
 
+  test("user type with ScalaReflection") {
+    val point = (new ExamplePoint(0.1, 0.2), new ExamplePoint(0.3, 0.4))
+    val schema = ScalaReflection.schemaFor[Tuple2[ExamplePoint, ExamplePoint]]
+      .dataType.asInstanceOf[StructType]
+    val attributeSeq = schema.toAttributes
+    val boundEncoder = encoderFor[Tuple2[ExamplePoint, ExamplePoint]]
+      .resolve(attributeSeq, outers).bind(attributeSeq)
+    assert(boundEncoder.fromRow(boundEncoder.toRow(point)) === point)
+  }
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala]: Just make this `productTest(("UDT", new ExamplePoint(0.1, 0.2)))`.  It always better to reuse existing infrastructure when possible.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 3 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Forgot to remove them.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala (hunk 4 above)
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala]: Thanks. I've updated it according to your comments.

**Comment by marmbrus** [comment]
Thanks, merging to master.

**Pull request closed without merging**